### ISSUE TITLE
Add lxc config device add disk --type=block for containers 

### ISSUE THREAD

**Issue by kamzar1** [lxc/lxd#10077]
Whilst a custom storage volume --type=block can be added to a VM as block device, it can not be attached to a container.
Also attaching a custom storage volume to a container as unix-block not possible due to missing source path.

This feature would make the procedure of custom volume handling uniform and consistent throughout instance types and storage volume types.

**Comment by stgraber** [comment]
This should be easy enough to sort out. This would effectively behave like a `unix-block` device and so would require `path` to be set so we know what path to put the block device on.

**Comment by cagriyildirimR** [comment]
Hi. I would like to work on this issue. If there's any further information I need to be aware, please let me know. Thank you.

**Comment by tomponline** [comment]
Thanks!

**Comment by cagriyildirimR** [comment]
Hi @stgraber, @tomponline. Before implementing this feature, I read through the documentation and experimented with LXD. 

Based on conversations here, including the forum, my understanding is that first we need to create a custom storage volume with block content type. We would then be able to add/attach this volume using either 

`lxc storage volume attach pool_name custom_volume_name instance_name path_name`

`lxc config device add instance_name custom_volume_name disk pool=pool_name source=source_name path=path_name`

However, both methods result in an error message stating that "Custom block volumes cannot be used on containers."

The validation function in the disk package checks the supplied configuration for correctness. Specifically, it checks that block volumes are only attached to virtual machine (VM) instances, not containers. If a block volume is attached to a container, the function will return an error message.

One question I had was why there is a restriction on using custom block volumes with containers in the first place. I couldn't find any information on this.

There is an alternative approach (as discussed in forum) which involves creating a block device on the host and then adding it to the container via a loop device using LXD. This somehow evades validation. This may be implemented for `lxc config device add disk --type=block` command. However, I don't believe this is what was intended for this feature. Could you provide me some insight?

**Comment by stgraber** [comment]
Removing this restriction is what this issue is about. The restriction should be lifted and the disk be attached to the container as a unix block device in /dev/

**Comment by stgraber** [comment]
With the fix for this issue applied, I should be able to do:
 - lxc storage volume create default foo --type block size=5GiB
 - lxc launch images:ubuntu/22.04 c1
 - lxc config device add c1 foo disk pool=default source=foo path=/dev/sda

And get a /dev/sda device in the container which is that custom block volume called foo.

**Comment by tomponline** [comment]
Out of interest what is your use case for this?

**Comment by markrattray** [comment]
Good morning @tomponline 

Not everything likes the underlying files system on filesystem volumes or performs better on another file system, so in some circumstances we need to be able to choose and block volume does this for containers. 

Docker is a good example with ZFS backed pools and I don't want to turn on blocks at the pool level. ArangoDB doesn't like BTRFS (not that I use BTRFS). A rare occurrence but they are out there. 

Although there is a work-around as linked in previous comments, it complicates automation, e.g.: Ansible. On my side, it's **not** urgent at all because the work-around works and I have no immediate need with Docker Swarm having to reside in VMs unfortunately.

Thanks
Mark

**Comment by tomponline** [comment]
For ZFS block mode and Docker you can use per-volume block mode now by the way:

```
lxc storage volume create zfs vol1 size=1GiB zfs.block_mode=true
lxc config device add c1 mydisk disk pool=myzfs source=vol1  path=/mnt
lxc exec c1 -- findmnt /mnt
TARGET SOURCE                              FSTYPE OPTIONS
/mnt   /dev/zvol/myzfs/custom/default_vol1 ext4   rw,relatime,discard,stripe=2
```

**Comment by tomponline** [comment]
> Good morning @tomponline
> 
> Not everything likes the underlying files system on filesystem volumes or performs better on another file system, so in some circumstances we need to be able to choose and block volume does this for containers.
> 
> Docker is a good example with ZFS backed pools and I don't want to turn on blocks at the pool level. ArangoDB doesn't like BTRFS (not that I use BTRFS). A rare occurrence but they are out there.
> 
> Although there is a work-around as linked in previous comments, it complicates automation, e.g.: Ansible. On my side, it's **not** urgent at all because the work-around works and I have no immediate need with Docker Swarm having to reside in VMs unfortunately.
> 
> Thanks Mark

The reason I ask is that attaching block disks to unprivileged containers is still going to be fairly restricted as unprivileged containers don't allow mounting. So if the application is using the block device directly then that is fine, but if the requirement is for attaching a block disk and then mounting it inside the container then you may be disapointed.

**Comment by markrattray** [comment]
**Re: ZFS Block Mode**
Thanks for that info, I'm a bit behind. In your example, how did it know to format it EXT4? I've done a quick search in documentation for this... I can only see some `volume.block.filesystem` in the [Specifications](https://discuss.linuxcontainers.org/t/lxd-zfs-block-mode/15872) doc. So we would have to issue a command like this to get EXT4?
```
lxc storage volume create zfs vol1 size=1GiB zfs.block_mode=true volume.block.filesystem=ext4
```
or would that now be 
```
... zfs.block.filesystem=ext4
```

**RE: Attaching block disk to unpriviledged containers**
So I was bind mounting the attached block disk formatted with EXT4 via fstab, for a few reasons:

- get Docker on a supported file system when the underlying storage pools are ZFS
- put all data to be backed up on a separate storage volume for snapshots of specific data
- being able to plug the volume/copy/snapshot into another instance if needed
- we traditionally have OS disks on cheaper SAS HDD RAID10, and data (clustered) on fast more expensive enterprise grade NVMe

```
# docker disk
#   - /mnt-acme/c1_disk00: where the block device is mounted
#   - c1_disk00: underlying block volume name : {container-name}_disk{instance}
# 
/mnt-acme/c1_disk00/var/lib/docker /var/lib/docker none auto,bind 0 0
/mnt-acme/c1_disk00/etc/docker /etc/docker none auto,bind 0 0
/mnt-acme/c1_disk00/some/other/important/data /some/other/important/data none auto,bind 0 0
```

**Comment by tomponline** [comment]
> or would that now be
> 
> ```
> ... zfs.block.filesystem=ext4
> ```

Yes thats correct. It defaults to ext4 if not specified.

**Comment by tomponline** [comment]
> > The reason I ask is that attaching block disks to unprivileged containers is still going to be fairly restricted as unprivileged containers don't allow mounting. So if the application is using the block device directly then that is fine, but if the requirement is for attaching a block disk and then mounting it inside the container then you may be disapointed.
> 
> This is fine - once this is implemented we should be able to allow mounting for the unprivileged containers with the following, if Im not mistaken
> 
> ```
> lxc config set c1 security.syscalls.intercept.mount.allowed=ext4 security.syscalls.intercept.mount=true
> ```

Yes, if you're happy with the risks involved (a corrupted or malicious superblock could crash the host kernel).

**Comment by allanrogerr** [comment]
> ```
> lxc storage volume create zfs vol1 size=1GiB zfs.block_mode=true volume.block.filesystem=ext4
> ```

@markrattray Are you sure, I was only able to get this to work with the following:
```
lxc storage volume create zfs vol1 size=1GiB zfs.block_mode=true block.filesystem=ext4
```

Do you know if there is a glossary of `configuration_options` out there? Otherwise I would just infer it from https://github.com/lxc/lxd/blob/master/lxd/storage/drivers/driver_zfs_volumes.go#L1463

**Comment by tomponline** [comment]
Ah yes, sorry, its block.filesystem.

See the tip at the bottom of https://linuxcontainers.org/lxd/docs/master/reference/storage_zfs/#configuration-options that links to https://linuxcontainers.org/lxd/docs/master/howto/storage_volumes/#storage-configure-vol-default

**Comment by tomponline** [comment]
Although block.filesystem is missing from https://linuxcontainers.org/lxd/docs/master/reference/storage_zfs/#storage-volume-configuration @ru-fu are you able to add it based on the one from https://linuxcontainers.org/lxd/docs/master/reference/storage_lvm/#storage-volume-configuration cc @monstermunchkin

**Comment by casperan** [comment]
usecase idea: would this also make it viable or at least possible to make block devices / unformatted filesystems in containers ? so it is possible to run a ceph cluster in containers using parts of disks using zfs volumes ? I understand that it is not possible to make pure gpt partitions available in containers, but with zfs and zvol and zfs now in containers, it becomes possible ?
